src/System.Net.Http/src/System/Net/Http/Managed/AuthenticateAndRedirectHandler.cs

```diff
@@ -0,0 +1,223 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http
+{
+    internal sealed partial class AuthenticateAndRedirectHandler : HttpMessageHandler
+    {
+        private readonly HttpMessageHandler _innerHandler;
+        private readonly bool _preAuthenticate;
+        private readonly ICredentials _credentials;
+        private readonly bool _allowRedirect;
+        private readonly int _maxAutomaticRedirections;
+        private AuthenticationHelper.DigestResponse _digestResponse;
+
+        public AuthenticateAndRedirectHandler(bool preAuthenticate, ICredentials credentials, bool allowRedirect, int maxAutomaticRedirections, HttpMessageHandler innerHandler)
+        {
+            Debug.Assert(innerHandler != null);
+
+            _preAuthenticate = preAuthenticate;
+            _credentials = credentials;
+            _allowRedirect = allowRedirect;
+
+            if (allowRedirect && maxAutomaticRedirections < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(maxAutomaticRedirections));
+            }
+
+            _maxAutomaticRedirections = maxAutomaticRedirections;
+            _innerHandler = innerHandler;
+        }
+
+        protected internal override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            HttpResponseMessage response;
+            uint redirectCount = 0;
+            bool useCredentialCache = false;
+            while (true)
+            {
+                response = await SendRequestAsync(request, useCredentialCache, cancellationToken).ConfigureAwait(false);
+
+                if (!RequestNeedsRedirect(response))
+                {
+                    break;
+                }
+
+                // Clear the authorization header, if the request requires redirect.
+                request.Headers.Authorization = null;
```

Comment:
>request.Headers.Authorization = null; [](start = 16, length = 37)

To David's point about mutating request.Headers, this one does seem potentially problematic.  It looks like the reason we're nulling this out is because the auth helpers code may have set one?  That's a) exposing auth-related information out to a caller that could be reading the HttpRequestMessage, b) adding allocations to store auth header classes into the request only to then just serialize them out, and c) mutating something that apparently we're supposed to minimize mutations of.  Given all that, is there an alternate way we could pass additional headers we want added to a request down to actually be written out?  That seems better.

Again, the previous auth code you added was already mutating this header (and ProxyAuthorization), so this isn't making things worse, but it'd be good to open an issue to track fixing this after this PR.

Reply:
> is there an alternate way we could pass additional headers we want added to a request down to actually be written out?

Not at the moment, will file an issue to work on it separately.